src/snowflake/snowpark/mock/_functions.py

```diff
@@ -495,45 +501,63 @@ def mock_to_time(
             if data is None:
                 res.append(None)
                 continue
-            if auto_detect and data.isnumeric():
-                res.append(
-                    datetime.datetime.utcfromtimestamp(
-                        process_numeric_time(data)
-                    ).time()
-                )
-            else:
-                # handle seconds fraction
-                data_parts = data.split(".")
-                if len(data_parts) == 2:
-                    # there is a part of seconds
-                    seconds_part = data_parts[1]
-                    # find the idx that the seconds part ends
-                    idx = 0
-                    while seconds_part[idx].isdigit():
-                        idx += 1
-                    # truncate to precision
-                    seconds_part = (
-                        seconds_part[: min(idx, fractional_seconds)]
-                        + seconds_part[idx:]
-                    )
-                    data = f"{data_parts[0]}.{seconds_part}"
-                res.append(
-                    (
-                        datetime.datetime.strptime(
-                            process_string_time_with_fractional_seconds(
-                                data, fractional_seconds
-                            ),
-                            time_fmt,
-                        )
-                        + datetime.timedelta(hours=hour_delta)
-                    ).time()
+            if auto_detect:
+                # handling non-standard string expr cases
+                if isinstance(column.sf_type.datatype, StringType) and data.isnumeric():
+                    # integer timestamp string '1234567890'
+                    res.append(convert_int_string_to_time(data))
+                    continue
+                # timestamp data
+                elif isinstance(column.sf_type.datatype, TimestampType):
+                    res.append(data.time())
+                    continue
+                # if the data doesn't fall into the above two categories, it goes to the string_expr handling
+                elif isinstance(column.sf_type.datatype, VariantType):
+                    """
+                    https://docs.snowflake.com/en/sql-reference/functions/to_time#arguments
+                    For variant_expr:
+                        If the variant contains a string in TIME format (‘HH:MI:SS’), a string conversion is performed.
+                        If the variant contains a string in INTEGER format, a string conversion is performed and the value is treated as the number of seconds since midnight (modulus 86400 if necessary).
+                        If the variant contains JSON null value, the output is NULL.
```

Comment:
This case also needs to handle when `data` is of `datetime` type, i.e. the first case in:
```
select to_time(current_time()::variant);
select to_time(current_time()::varchar);
```

Reply:
good point, code updated to handle variant type with value being time, also added a test for it